### PR TITLE
Allow custom PR ops with sparse pools

### PR DESCRIPTION
**Pull request by pdmullen**
## PR Summary
`AddSparsePool` copies the relevant `Metadata` for all entries in the pool.  Calling the `Metadata` constructor upon copying (unintentionally) overwrites any custom P/R operations that were previously enrolled via `m.RegisterRefinementOps`.

My changeset adds the default PR ops as an optional argument to the `Metadata` constructor.  When `IsRefined()`, this default argument is used to set the `refinement_funcs_` private member.  

With this change, when copying `Metadata` for all entries in the `SparsePool`, we can just pass `shared_metadata.GetRefinementFunctions()` to the argument list, resulting in the desired behavior.  

My changeset feels somewhat hacky, and invokes unnecessary `RegisterOps` calls due to the default argument.  Please be harsh... if anything, let this PR bring attention to the issue and a possible resolution, even if we eventually take a wildly different approach to solving it.  

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] Change is breaking (API, behavior, ...)
  - [ ] Change is *additionally* added to CHANGELOG.md in the breaking section
  - [ ] PR is marked as breaking
  - [ ] Short summary API changes at the top of the PR (plus optionally with an automated update/fix script)
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] Docs build
- [x] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

File: src/interface/metadata.hpp [hunk 1]
```diff
@@ -548,7 +552,6 @@ class Metadata {
 
   // Refinement stuff
   const refinement::RefinementFunctions_t &GetRefinementFunctions() const {
-    PARTHENON_REQUIRE_THROWS(IsRefined(), "Variable must be registered for refinement");
```
Comment (Yurlungur) [inline src/interface/metadata.hpp:551]: What is the reason for this change?

**Review by Yurlungur — CHANGES_REQUESTED** [review]
Good catch, thanks for finding this! Big picture, I wonder if a more maintainable solution is to add a copy constructor and assignment operator... probably enough to do
```C++
Metadata::Metadata(const Metadata &other) = default;
friend Metadata &operator=(const Metadata &other) = default;
```
and then use assignment when copying in the sparse pool?

We could also add a test in the test suite that the copy constructor works, and in sparse pool that custom prolong/restrict works?

File: src/interface/metadata.hpp (hunk 1 above)
Comment (pdmullen) [reply src/interface/metadata.hpp:551]: We are calling `GetRefinementFunctions()` here: 

https://github.com/parthenon-hpc-lab/parthenon/pull/992/files#diff-e142fce743a6bbfa5d718f415ca450fd5abf0694750a16562c64f7b5e392d5e6R62

With the current design, we need to make this call even for fields that are *not* `IsRefined()`.  In this scenario, we are passing an unset `refinement_funcs_`, which is never actually used because there is an internal check on `IsRefined()` inside the `Metadata` constructor, followed by an additional `IsValid()`.

Comment (Yurlungur) [reply src/interface/metadata.hpp:551]: Oh I see. So long as it's still safe somewhere, I think this is fine. Tests will fail though (because I think this is a test case.)

**Comment by pdmullen** [comment]
Added a few tests that `refinement_funcs_` contains the expected entries after adding a `SparsePool`.  The newly added test fails under current `develop` (as it should).

File: tst/unit/test_state_descriptor.cpp [hunk 2]
```diff
@@ -383,16 +383,16 @@ TEST_CASE("Test dependency resolution in StateDescriptor", "[StateDescriptor]")
       }
     }
 
-    WHEN("We register a dense variable custom prolongation/restriction") {
+    WHEN("We register a dense variable with default prolongation/restriction") {
```
Comment (Yurlungur) [inline tst/unit/test_state_descriptor.cpp:386]: These are still custom right? Not default?

**Review by Yurlungur — APPROVED** [review]
LGTM. Thanks for adding the test coverage!

File: tst/unit/test_state_descriptor.cpp (hunk 2 above)
Comment (pdmullen) [reply tst/unit/test_state_descriptor.cpp:386]: I don't think so... only the `SparsePool` has enrolled custom P/R ops via `RegisterRefinementOps`.  The dense field just uses the default ones.

Comment (Yurlungur) [reply tst/unit/test_state_descriptor.cpp:386]: Oh yes you're right. :+1:

**Review by brryan — APPROVED** [review]
LGTM, I am not offended by the default call to RegisterOps. Only question is whether metadata is the right place for default values of register ops to live (maybe it is). If not, one could do something like
```c++
using DefaultRegisterOps = RegisterOps<refinement_ops::ProlongateSharedMinMod, refinement_ops::RestrictAverage>;
```
in the refinement functions file.

**Comment by pdmullen** [comment]
> LGTM, I am not offended by the default call to RegisterOps. Only question is whether metadata is the right place for default values of register ops to live (maybe it is). If not, one could do something like
> 
> ```c++
> using DefaultRegisterOps = RegisterOps<refinement_ops::ProlongateSharedMinMod, refinement_ops::RestrictAverage>;
> ```
> 
> in the refinement functions file.

I see your point.  I could imagine a scenario where one wants to handle different fields via different P/R ops, therefore, I think that `Metadata` is a natural place for this to live, however, others may disagree.
